File: src/sub_converter/services/subscription.py
"""Fetch, decode, and split raw subscription data."""

from __future__ import annotations

import base64
import binascii
import re
import ssl
from urllib.request import Request, urlopen

from sub_converter.utils import pad_base64


def fetch_subscription(url: str, timeout: int) -> bytes:
    request = Request(
        url,
        headers={
            "User-Agent": "sub-converter/1.0 (+https://local.script)",
            "Accept": "*/*",
        },
    )
    context = ssl.create_default_context()
    with urlopen(request, timeout=timeout, context=context) as response:
        return response.read()


def try_decode_base64_text(raw: bytes) -> str | None:
    text = raw.decode("utf-8", errors="ignore").strip()
    if not text:
        return ""

    candidate = pad_base64(text)
    if not re.fullmatch(r"[A-Za-z0-9+/=]+", candidate):
        return None

    try:
        decoded = base64.b64decode(candidate, validate=False)
    except (binascii.Error, ValueError):
        return None

    decoded_text = decoded.decode("utf-8", errors="ignore").strip()
    if not decoded_text:
        return None

    known_prefixes = (
        "ss://", "ssr://", "vmess://", "vless://", "trojan://",
        "socks://", "http://", "https://", "hy2://", "hysteria2://",
        "tuic://", "wireguard://",
    )
    if decoded_text.startswith(known_prefixes) or "\n" in decoded_text:
        return decoded_text

    return None


def subscription_text(raw: bytes) -> tuple[str, str]:
    decoded = try_decode_base64_text(raw)
    if decoded is not None:
        return decoded, "base64"
    return raw.decode("utf-8", errors="ignore"), "plain"


def split_entries(text: str) -> list[str]:
    entries: list[str] = []
    for line in text.splitlines():
        item = line.strip()
        if not item or item.startswith("#"):
            continue
        entries.append(item)
    return entries

File: worker/src/services/subscription.ts
/**
 * Utilities for subscription handling
 */

export function padBase64(str: string): string {
  const trimmed = str.trim().replace(/\s+/g, '').replace(/-/g, '+').replace(/_/g, '/');
  const remainder = trimmed.length % 4;
  if (remainder === 0) return trimmed;
  return trimmed + '='.repeat(4 - remainder);
}

export function isValidBase64(str: string): boolean {
  try {
    const padded = padBase64(str);
    if (!/^[A-Za-z0-9+/=]+$/.test(padded)) return false;
    atob(padded);
    return true;
  } catch {
    return false;
  }
}

export function tryDecodeBase64(raw: string): string | null {
  if (!raw || !raw.trim()) return '';

  const candidate = padBase64(raw.trim());
  if (!/^[A-Za-z0-9+/=]+$/.test(candidate)) {
    return null;
  }

  try {
    const decoded = atob(candidate);
    const text = decoded.trim();
    if (!text) return null;

    const knownPrefixes = [
      'ss://', 'ssr://', 'vmess://', 'vless://', 'trojan://',
      'socks://', 'http://', 'https://', 'hy2://', 'hysteria2://',
      'tuic://', 'wireguard://',
    ];

    if (knownPrefixes.some(p => text.startsWith(p)) || text.includes('\n')) {
      return text;
    }
  } catch {
    return null;
  }

  return null;
}

export async function fetchSubscription(url: string, timeout: number = 10000): Promise<string> {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeout);

  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'sub-converter-worker/1.0 (+https://workers.dev)',
        'Accept': '*/*',
      },
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const buffer = await response.arrayBuffer();
    const bytes = new Uint8Array(buffer);
    const text = new TextDecoder('utf-8').decode(bytes);

    return text;
  } finally {
    clearTimeout(timeoutId);
  }
}

export function subscriptionText(raw: string): [string, string] {
  const decoded = tryDecodeBase64(raw);
  if (decoded !== null) {
    return [decoded, 'base64'];
  }
  return [raw, 'plain'];
}

export function splitEntries(text: string): string[] {
  const entries: string[] = [];
  for (const line of text.split('\n')) {
    const item = line.trim();
    if (item && !item.startsWith('#')) {
      entries.push(item);
    }
  }
  return entries;
}
